squint.c

```diff
@@ -0,0 +1,1611 @@
+            match = (*target & 0xf0000000);
+            while(1) {
+               /* calc sign extended branch addr */
+               tmp = (*target & 0x00ffffff) |
+                     ((*target & 0x00800000) ? 0xff000000 : 0);
+               target += 2 + tmp;
+               if (*target == 0xe3500000) { /* cmp r0, #0 */
+                  target = active_inst(target, 1);
+               }
+               if ((*target & 0xff000000) == 0x0a000000 || /* beq */
+                   (*target & 0xff000000) == 0x1a000000) { /* bne */
+                  if ((*target & 0xf0000000) != match) {
+                     ++target;
+                     if ((*target & 0xff000000) != 0xea000000) { /* b */
+                        break;
+                     }
+                  }
+               }
+               else if ((*target & 0xff000000) != 0xea000000) { /* b */
+                  break;
+               }
+            }
+            *branchInst = (*branchInst & 0xff000000) |
+                          ((target - branchInst - 2) & 0xffffff);
+         }
+      }
+   }
+}
+
+/* Now that we have simplified all final branch targets, we */
+/* will simplify and compress out extraneous comparison opcodes. */
+void simplifyBranch4(int *funcBegin, int *funcEnd)
+{
+   int *scan;
+   for (scan = funcBegin; scan < funcEnd; ++scan) {
+      scan = skip_nop(scan, 1);
+
+      if (*scan == 0xe3500000) { /* cmp r0, #0 */
+         int *scanm1, *scanm2, *scanm3, *scanp1;
+
+         /* simplify cmp blocks */
+         scanm1 = active_inst(scan,   -1);
+         scanm2 = active_inst(scanm1, -1);
+         scanm3 = active_inst(scanm2, -1);
+         if ((*scanm3 & 0xf3f0f000) == 0xe1500000)  { /* cmp rX, r0 */
+            scanp1 = active_inst(scan, 1);
+            int match = ((*scanp1 & 0xf0000000) == 0) ? 0 /* eq */ : 1 /* ne */;
+            if ((*scanm1 & 1) == match) {
+               *scanp1 = (*scanp1 & 0x0fffffff) | (*scanm1 & 0xf0000000);
+            }
+            else {
+               *scanp1 = (*scanp1 & 0x0fffffff) | (*scanm2 & 0xf0000000);
+            }
+            *scanm2 = NOP;
+            *scanm1 = NOP;
+            *scan   = NOP;
+         }
+      }
+   }
+}
+
+/* "bcond target1 ; b target2 ; taget1: <inst>;" -> "bnotcond target2;" */
+/* todo: may want to check for chains of intervening nops here */
+void simplifyBranch5(int *funcBegin, int *funcEnd)
+{
+   int *scan;
+   for (scan = funcBegin; scan < funcEnd; ++scan) {
+      scan = skip_nop(scan, 1);
+      if ((*scan & 0x0fffffff) == 0x0a000000 &&  // cond jump past next inst
+          ((*scan >> 28) &0xf) < 0xe &&
+          (scan[1] & 0xff000000) == 0xea000000) {
+         scan[1] = (scan[0] & 0xff000000) | (scan[1] & 0x00ffffff);
+         scan[1] = scan[1] ^ (int) 0x10000000;
+         scan[0] = NOP;
+      }
+   }
+}
+
+
+/* optimize branch-specific code sequences */
+void simplifyBranch(int *funcBegin, int *funcEnd) {
```

Comment:
simplify_branch

I see other occurrences use `apply_peepholes1`, thus I suggest to change these names for consistency.